src/chanjo2/meta/handle_report_contents.py

```diff
@@ -28,24 +45,84 @@ def get_report_data(query: ReportQuery, session: Session) -> Dict:
+        (sample.name, D4File(sample.coverage_file_path)) for sample in query.samples
+    ]
+    genes: List[SQLGene] = get_genes(
+        db=session,
+        build=query.build,
+        ensembl_ids=query.ensembl_gene_ids,
+        hgnc_ids=query.hgnc_gene_ids,
+        hgnc_symbols=query.hgnc_gene_symbols,
+        limit=None,
+    )
+    samples_coverage_stats: Dict[str, List[GeneCoverage]] = {
+        sample: get_sample_interval_coverage(
+            db=session,
+            d4_file=d4_file,
+            genes=genes,
+            interval_type=INTERVAL_TYPE_SQL_TYPE[query.interval_type],
+            completeness_thresholds=query.completeness_thresholds,
+        )
+        for sample, d4_file in samples_d4_files
+    }
 
     data: Dicr = {
         "levels": get_ordered_levels(threshold_levels=query.completeness_thresholds),
         "extras": {
             "panel_name": query.panel_name,
             "default_level": query.default_level,
         },
-        "sex_rows": get_report_sex_rows(query.samples),
+        "sex_rows": get_report_sex_rows(
+            samples=query.samples, samples_d4_files=samples_d4_files
+        ),
+        "completeness_rows": get_report_completeness_rows(
+            samples_coverage_stats=samples_coverage_stats,
+            levels=query.completeness_thresholds,
+        ),
     }
     return data
 
 
-def get_report_sex_rows(samples: List[ReportQuerySample]) -> List[Dict]:
+def get_report_completeness_rows(
+    samples_coverage_stats: Dict[str, List[GeneCoverage]], levels: List[int]
+):
+    """Create and return the contents for the samples' coverage completeness rows in the coverage report"""
+
+    completeness_rows: List[str, Dict[str, float]] = []
+
+    for sample, interval_stats in samples_coverage_stats.items():
+        sample_stats: Dict[str, List[float]] = {"mean_coverage": []}
+        for level in levels:
+            sample_stats[f"completeness_{level}"]: List[float] = []
+
+        for interval in interval_stats:
+            sample_stats["mean_coverage"].append(interval.mean_coverage)
+            for level in levels:
+                sample_stats[f"completeness_{level}"].append(
+                    interval.completeness[level]
+                )
+
+        completeness_row: Dict[str, float] = {}
+        for item, values in sample_stats.items():
+            completeness_row[item] = mean(values) if values else 0
+
+        completeness_rows.append((sample, completeness_row))
+
+    return completeness_rows
+
+
+def get_report_sex_rows(
+    samples: List[ReportQuerySample], samples_d4_files: List[Tuple[str, D4File]]
+) -> List[Dict]:
     """Create and return the contents for the sample sex lines in the coverage report."""
     sample_sex_rows: D4FileList = []
     for sample in samples:
-        sample_d4: D4File = D4File(sample.coverage_file_path)
-        sample_sex_metrics: Dict = get_samples_sex_metrics(d4_file=sample_d4)
+        for _, d4_file in samples_d4_files:
```

Comment:
```suggestion
        for identifier, d4_file in samples_d4_files:
```
? You are using the "_" so that feels incorrect.